### PR TITLE
Add DH key exchange to fips provider

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [ ] documentation is added or updated
- [X] tests are added or updated

### CODE REVIEW

File: crypto/dh/build.info [hunk 1]
```diff
@@ -1,5 +1,10 @@
...
-        dh_asn1.c dh_gen.c dh_key.c dh_lib.c dh_check.c dh_err.c dh_depr.c \
+
+$COMMON=dh_lib.c dh_key.c
+
+SOURCE[../../libcrypto]=$COMMON\
+        dh_asn1.c dh_gen.c dh_check.c dh_err.c dh_depr.c \
         dh_ameth.c dh_pmeth.c dh_prn.c dh_rfc5114.c dh_kdf.c dh_meth.c \
         dh_rfc7919.c
+
+SOURCE[../../providers/libfips.a]=$COMMON
```
Comment (paulidale) [inline crypto/dh/build.info]: New at end of file.

Comment (slontis) [reply crypto/dh/build.info]: Fixed

File: providers/fips/fipsprov.c [hunk 2]
```diff
@@ -164,6 +164,193 @@ static int rawnative_fromhex(const char *hex_data[],
...
+{
+    int ret = 0;
+    unsigned char *data = NULL;
+    BIGNUM *bn = NULL;
+    int i, slen, datalen, sz;
+    char *str = NULL;
+
+    for (slen = 0, i = 0; hex_data[i] != NULL; ++i)
+        slen += strlen(hex_data[i]);
+    str = OPENSSL_zalloc(slen + 1);
```
Comment (levitte) [inline providers/fips/fipsprov.c]: Free it somewhere?

Comment (slontis) [reply providers/fips/fipsprov.c]: The caller needed to free the secret.

**Comment by slontis** [comment]
@levitte - you want me to remove the ex_data from this PR i assume?.

**Comment by levitte** [comment]
Yes please

**Comment by slontis** [comment]
As discussed - the fips provider temporary test code to load BIGNUM's has been changed to use the param_builder.. This code will be moved into self tests in a future PR.

File: crypto/dh/dh_key.c [hunk 3]
```diff
@@ -11,29 +11,94 @@
...
+    }
+/* TODO(3.0) : Solve in a PR related to Key validation for DH */
+#ifndef FIPS_MODE
+    if (!DH_check_pub_key(dh, pub_key, &check_result) || check_result) {
+        DHerr(0, DH_R_INVALID_PUBKEY);
+        goto err;
+    }
+#endif
+    if (!dh->
+        meth->bn_mod_exp(dh, tmp, pub_key, dh->priv_key, dh->p, ctx, mont)) {
```
Comment (paulidale) [inline crypto/dh/dh_key.c]: Is there a nicer line break here?

File: providers/fips/fipsprov.c [hunk 4]
```diff
@@ -223,14 +239,237 @@ static int dummy_evp_call(void *provctx)
...
+    static const char *dh_secret_exptd_hex[] = {
+        "08ff33bb2ecff49a7d4a7912aeb1bb6ab511641b4a76770c8cc1bcc2"
+        "33343dfe700d11813d2c9ed23b211ca9e8786921edca283c68b16153"
+        "fa01e91ab82c90ddab4a95816770a98710e14c92ab83b6e46e1e426e"
+        "e852430d6187daa3720a6bcd73235c6b0f941f3364f50420551a4bfe",
+        "afe2bc438505a59a4a40daca7a895a73db575c74c13a23ad8832957d"
+        "582d38f0a6165fb0d7e9b8799e42fd3220e332e98185a0c9429757b2"
+        "d0d02c17dbaa1ff6ed93d7e73e241eaed90caf394d2bc6570f18c81f"
+        "2be5d01a2ca99ff142b5d963f9f500325e7556f95849b3ffc7479486",
+        "be1d4596a3106bd5cb4f61c57ec5f100fb7a0c82a10b82526a97d1d9",
```
Comment (paulidale) [inline providers/fips/fipsprov.c]: Drop the comma here?  It's four lines concatenating previously but one here which isn't consistent.

File: providers/fips/fipsprov.c [hunk 5]
```diff
@@ -122,65 +123,80 @@ static OSSL_PARAM core_params[] =
...
 err:
+    OPENSSL_free(buf);
     BN_free(bn);
-    OPENSSL_free(data);
     return ret;
 }
 
-/* TODO(3.0): To be removed */
-static int dummy_evp_call(void *provctx)
+#ifndef OPENSSL_NO_DSA
```
Comment (paulidale) [inline providers/fips/fipsprov.c]: A wider question: should the FIPS provider include conditional checks for `OPENSSL_NO_DSA` and other macros?

They are beneficial for 3.0 (and likely validations beyond that).  They complicate the source code.  They could be useful to third parties.

File: providers/implementations/keymgmt/dh_kmgmt.c [hunk 6]
```diff
@@ -15,6 +15,7 @@
 #include "internal/param_build.h"
 #include "prov/implementations.h"
 #include "prov/providercommon.h"
+#include "prov/provider_ctx.h"
```
Comment (levitte) [inline providers/implementations/keymgmt/dh_kmgmt.c]: ?

File: providers/fips/fipsprov.c (hunk 5 above)
Comment (mattcaswell) [reply providers/fips/fipsprov.c]: Yes. If someone compiles with "no-dsa" then that is what they should get - including in the fips module.

File: providers/fips/fipsprov.c (hunk 4 above)
Comment (richsalz) [reply providers/fips/fipsprov.c]: There's a limit on static stringth length, so perhaps put a comma on all lines and glue them together as is done elsewhere?

File: providers/implementations/keymgmt/dh_kmgmt.c (hunk 6 above)
Comment (slontis) [reply providers/implementations/keymgmt/dh_kmgmt.c]: It was for the libctx - removed now,

**Comment by slontis** [comment]
Applied fixups and rebased.

**Comment by slontis** [comment]
ping

File: providers/fips/fipsprov.c [hunk 7]
```diff
@@ -223,14 +239,237 @@ static int dummy_evp_call(void *provctx)
...
+    EVP_PKEY_free(pkey);
+    EVP_PKEY_free(peerkey);
+    EVP_PKEY_CTX_free(kactx);
+    EVP_PKEY_CTX_free(dctx);
+    return ret;
+}
+#endif /* OPENSSL_NO_DH */
+
+/* TODO(3.0): To be removed */
+static int dummy_evp_call(void *provctx)
```
Comment (mattcaswell) [inline providers/fips/fipsprov.c]: Side note: I wonder how much longer we need to retain this.

Comment (slontis) [reply providers/fips/fipsprov.c]: Will remove when all the self tests are added..

**Comment by slontis** [comment]
Thanks.. merged to master.

**Pull request closed without merging**